src/keyboard/plugins/character.ts

```diff
@@ -147,12 +147,65 @@ export const keypressBehavior: behaviorPlugin[] = [
         },
       })
 
-      const appliedValue = getValue(element)
-      if (appliedValue === newValue) {
-        state.carryValue = undefined
-      } else {
-        state.carryValue = newValue
-      }
+      setCarryValue(element, state, newValue)
+    },
+  },
+  {
+    matches: (keyDef, element) =>
+      keyDef.key?.length === 1 &&
+      isElementType(element, 'input', {type: 'email', readOnly: false}),
+    handle: (keyDef, element, options, state) => {
+      const oldValue =
+        state.carryValue ?? getValue(element) ?? /* istanbul ignore next */ ''
+
+      const {newValue, newSelectionStart} = calculateNewValue(
+        keyDef.key as string,
+        element as HTMLElement,
+        oldValue,
+      )
+
+      fireInputEvent(element as HTMLElement, {
+        newValue,
+        newSelectionStart,
+        eventOverrides: {
+          data: keyDef.key,
+          inputType: 'insertText',
+        },
+      })
+
+      setCarryValue(element, state, newValue)
+    },
+  },
+  {
+    matches: (keyDef, element) =>
+      keyDef.key?.length === 1 && isElementType(element, 'input', {type: 'url', readOnly: false}
+      ),
+    handle: (keyDef, element, options, state) => {
+      const oldValue =
+        state.carryValue ?? getValue(element) ?? /* istanbul ignore next */ ''
+
+      const selectionRange = oldValue.endsWith(" ") ? {
+        selectionStart: null,
+        selectionEnd: null
+      } : undefined
+
+      const {newValue, newSelectionStart} = calculateNewValue(
+        keyDef.key as string,
+        element as HTMLElement,
+        oldValue,
+        selectionRange
```

Comment:
Am I interpreting this correctly that you set `selectionRange` to force `calculateNewValue` to not use `input.selectionStart` although it exists on this type of element?

Doesn't forcing the `selectionRange` here lead to diverging from other behavior implementations like `Backspace`?
`userEvent.type(input, 'foo {Backspace}bar')` => `fo bar` ?

Reply:
Ahhh yeah you are interpreting that correctly, that's my bad!!

I added that to handle the case for the element type being ur. This is due to getSelectionRange using the HTML element to determine the length and the input url element ignores the space.

Open to any suggestions on how you would handle this?

I thought of changing the beginning of getSelectionRange to the below. But I'm not convinced I like it 😂 

```
export function getSelectionRange(
  element: Element,
  value?: string
): {
  selectionStart: number | null
  selectionEnd: number | null
} {
  if (hasSelectionSupport(element)) {
    if (value && value.endsWith(" ") && element.type === 'url') {
      return {
        selectionStart: value.length,
        selectionEnd: value.length,
      }
    }

    return {
      selectionStart: element.selectionStart,
      selectionEnd: element.selectionEnd,
    }
  }
```

Reply:
Remember space at the end is not the only possible display value diverging from value property.
Also I don't like specific workarounds being scattered across codebase.
Maybe this should wait for #714

Reply:
Yeah, that's a good point. I'll close this pull request and have a look at a later point once #714 is resolved